src/windows/osd.c

```diff
@@ -402,64 +402,109 @@ ssize_t sendmsg(SOCKET sd, struct msghdr *msg, int flags)
 	goto fn_complete;
 }
 
+void sort_list_of_adapters(PIP_ADAPTER_ADDRESSES head)
+{
+	PIP_ADAPTER_ADDRESSES ptr = head, tmp = NULL, prev = NULL;
+	bool flag = false;
+	if (head) {
+		do {
+			flag = false;
+			ptr = head;
+			while (ptr->Next) {
+				if (ptr->Ipv4Metric < ptr->Next->Ipv4Metric) {
+					if (ptr == head) {
+						tmp = ptr;
+						ptr = tmp->Next;
+						tmp->Next = ptr->Next;
+						ptr->Next = tmp;
+						head = ptr;
+						flag = true;
+					} else {
+						tmp = ptr;
+						ptr = tmp->Next;
+						tmp->Next = ptr->Next;
+						ptr->Next = tmp;
+						prev->Next = ptr;
+						flag = true;
+					}
+				}
+				prev = ptr;
+				ptr = ptr->Next;
+			}
+		} while (flag);
+	}
+}
+
 int getifaddrs(struct ifaddrs **ifap)
 {
-	DWORD i;
-	MIB_IPADDRTABLE _iptbl;
-	MIB_IPADDRTABLE *iptbl = &_iptbl;
-	ULONG ips = 1;
-	ULONG res = GetIpAddrTable(iptbl, &ips, 0);
+	DWORD res, i = 0;
+	DWORD size;
+	PIP_ADAPTER_ADDRESSES adapter_addresses, aa;
+	PIP_ADAPTER_UNICAST_ADDRESS ua;
 	int ret = -1;
 	struct ifaddrs *head = NULL;
 
-	assert(ifap);
-
-	if (res == ERROR_INSUFFICIENT_BUFFER) {
-		iptbl = malloc(ips);
-		if (!iptbl)
-			goto failed_no_mem;
-		res = GetIpAddrTable(iptbl, &ips, 0);
-		if (res != NO_ERROR)
-			goto failed_get_addr;
-	} else if (res != NO_ERROR) {
+	res = GetAdaptersAddresses(AF_UNSPEC, GAA_FLAG_INCLUDE_PREFIX, NULL, NULL, &size);
+	if (res != ERROR_BUFFER_OVERFLOW) {
+		goto failed_no_mem;
+	}
+	adapter_addresses = (PIP_ADAPTER_ADDRESSES)malloc(size);
+	res = GetAdaptersAddresses(AF_UNSPEC, GAA_FLAG_INCLUDE_PREFIX, NULL, adapter_addresses, &size);
+	if (res != ERROR_SUCCESS) {
+		free(adapter_addresses);
```

Comment:
Move this free() call into the error path, so it gets called in all failure cases.  If we go through the failed_cant_allocate: label, the adapter_addresses are being leaked.

Reply:
@shefty Done.